### ISSUE TITLE
[video_player] Add possibility to start video playback from certain duration

### ISSUE THREAD

**Issue by IlyaMax** [flutter/flutter#97198]
## Use case

I need to start video playback from certain duration.

## Proposal

Add duration parameter to `initialize` method

**Comment by maheshmnj** [comment]
Hi @IlyaMax, Thanks for filing the issue. You could use `seekTo` which allows switching video to any duration. Please take a look at the below code sample which starts the video at 15th second.

> _controller.seekTo(const Duration(seconds: 15));

<details>
<summary>code sample</summary>

```dart
import 'package:video_player/video_player.dart';
import 'package:flutter/material.dart';

void main() => runApp(VideoApp());

class VideoApp extends StatefulWidget {
  @override
  _VideoAppState createState() => _VideoAppState();
}

class _VideoAppState extends State<VideoApp> {
  late VideoPlayerController _controller;

  @override
  void initState() {
    super.initState();
    _controller = VideoPlayerController.network(
        'https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ForBiggerFun.mp4')
      ..initialize().then((_) {
        _controller.seekTo(const Duration(seconds: 15));
        setState(() {});
      });
    _controller.play();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Video Demo',
      home: Scaffold(
        body: Center(
          child: _controller.value.isInitialized
              ? AspectRatio(
                  aspectRatio: _controller.value.aspectRatio,
                  child: VideoPlayer(_controller),
                )
              : Container(),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: () {
            setState(() {
              _controller.value.isPlaying
                  ? _controller.pause()
                  : _controller.play();
            });
          },
          child: Icon(
            _controller.value.isPlaying ? Icons.pause : Icons.play_arrow,
          ),
        ),
      ),
    );
  }

  @override
  void dispose() {
    super.dispose();
    _controller.dispose();
  }
}

```
</details>

**Comment by IlyaMax** [comment]
@maheshmnj when I tried to do something like that last time I had seen first frame but I think this shouldn't work like this. Are you sure your code sample doesn't have this behavior ok both android and ios?

**Comment by maheshmnj** [comment]
Thanks for the info @IlyaMax, I do see an initial frame with the above approach verified on Android.
Labeling this issue as a valid feature request.

**Comment by stuartmorgan** [comment]
> Add `duration` parameter to `initialize` method with optional value `Duration.zero`

Adding initialization parameters for things that change over time has generally been a source of confusion in other plugin controllers. (I assume you mean `position`, not `duration`, given the description.)

If the issue is the initial frame, perhaps a better solution would be a parameter that prevents drawing until the first `play` or `seekTo` call.

**Comment by IlyaMax** [comment]
Yes, position. What you mean by "source of confusion"? Can you give examples of such cases? You mean to force users of plugin to call `play` or `seekTo` after initialising or allow to call `seekTo` before player is initialized to make it work(is it possible?)?

**Comment by stuartmorgan** [comment]
> What you mean by "source of confusion"?

The usual pattern in Flutter is you change things by rebuilding then with different parameters. Plugin controllers like this one map to native objects where that's not the model, and instead everything is based on mutating persistent state.

Providing constructor arguments that correspond to *mutating* state values on the native side have very unclear semantics. E.g., if I have a video player with a `duration` value, what happens if I rebuild after three user has started playing the video? Does it reset the position? If so, that's likely unexpected. But what if I try to do the standard Flutter thing and rebuild with a different value? Does that change the position? If *not* that's unexpected.

> Can you give examples of such cases?

The initial URL parameter in the webview plugin has caused this kind of confusion for many people.



> You mean to force users of plugin to call `play` or `seekTo` after initialising

I don't understand what you mean by "force". You want to seek in the video. There's a method for seeking. Where is the "forcing" happening here?

**Comment by IlyaMax** [comment]
> Providing constructor arguments that correspond to mutating state values on the native side have very unclear semantics. E.g., if I have a video player with a duration value, what happens if I rebuild after the user has started playing the video? Does it reset the position? If so, that's likely unexpected. But what if I try to do the standard Flutter thing and rebuild with a different value? Does that change the position? If not that's unexpected.

You mean I need to beware of this case? That it can lead to unexpected behavior?
`controller.initialize(initialPosition:Duration(...)).then((_)=>controller.seekTo(Duration(...));`

> I don't understand what you mean by "force". You want to seek in the video. There's a method for seeking. Where is the "forcing" happening here?

You said 

> If the issue is the initial frame, perhaps a better solution would be a parameter that prevents drawing until the first play or seekTo call.

`controller.initialize(preventDrawingUntilSeekTo:true).then((_)=>controller.seekTo(Duration(...));`
I mean if I enable `preventDrawingUntilSeekTo` option I force users to call `seekTo`. Do I understand correctly your idea?

**Comment by IlyaMax** [comment]
Or maybe it would be better to do like this 
```dart
controller.initialize(initialPosition:Duration(...))
```
And inside `initialize` call `seekTo` and prevent drawing until seeking is completed

**Comment by stuartmorgan** [comment]
Sorry, I misread the proposal originally; I thought you wanted to make this a new *constructor* argument.

Looking at the code again, I don't think there's any reason we couldn't just make calling `seekTo` work before `initialize`, by storing the value for using during creation.
